src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,684 @@
+
+    // The name of the KDF algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final KDFParameters kdfParameters;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    // This lock provides mutual exclusion, preventing multiple threads from
+    // concurrently initializing the same instance (delayed provider selection)
+    // in a way which would corrupt the internal state.
+    private final Object lock = new Object();
+
+
+    // Instantiates a {@code KDF} object. This constructor is called when a
+    // provider is supplied to {@code getInstance}.
+    //
+    // @param delegate the delegate
+    // @param algorithm the algorithm
+    // @param kdfParameters the parameters
+    private KDF(Delegate delegate, String algorithm, KDFParameters kdfParameters) {
+        this.pairOfSpiAndProv = delegate;
+        this.algorithm = algorithm;
+        // note that the parameters are being passed to the impl in getInstance
+        this.kdfParameters = kdfParameters;
+        serviceIterator = null;
+    }
+
+    // Instantiates a {@code KDF} object. This constructor is called when a
+    // provider is not supplied to {@code getInstance}.
+    //
+    // @param firstPairOfSpiAndProv the delegate
+    // @param t the service iterator
+    // @param algorithm the algorithm
+    // @param kdfParameters the algorithm parameters
+    private KDF(Delegate firstPairOfSpiAndProv, Iterator<Service> t, String algorithm,
+                KDFParameters kdfParameters) {
+        this.firstPairOfSpiAndProv = firstPairOfSpiAndProv;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        useFirstSpi();
+        return pairOfSpiAndProv.provider().getName();
+    }
+
+    /**
+     * Returns the {@code KDFParameters} used with this {@code KDF} object.
+     * <p>
+     * The returned parameters may be the same that were used to initialize
+     * this {@code KDF} object, or may contain additional default or
+     * random parameter values used by the underlying KDF algorithm.
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
```

Comment:
Why see DPS? The paragraph above has nothing to do with it.

Reply:
Oh, sorry, I don't why I didn't see the `@see` in `getProviderName`.

If you need to refer to the reader to this information, please add it to `deriveData` and `deriveKey` as well.